File: src/main/java/org/jukeboxmc/item/behavior/ItemEgg.java
package org.jukeboxmc.item.behavior;

import com.nukkitx.protocol.bedrock.data.SoundEvent;
import org.jukeboxmc.entity.Entity;
import org.jukeboxmc.entity.EntityType;
import org.jukeboxmc.entity.projectile.EntityEgg;
import org.jukeboxmc.event.entity.ProjectileLaunchEvent;
import org.jukeboxmc.item.Item;
import org.jukeboxmc.item.ItemType;
import org.jukeboxmc.math.Vector;
import org.jukeboxmc.player.GameMode;
import org.jukeboxmc.player.Player;
import org.jukeboxmc.util.Identifier;

/**
 * @author LucGamesYT
 * @version 1.0
 */
public class ItemEgg extends Item {

    public ItemEgg( Identifier identifier ) {
        super( identifier );
    }

    public ItemEgg( ItemType itemType ) {
        super( itemType );
    }

    @Override
    public boolean useInAir( Player player, Vector clickVector ) {
        EntityEgg entityEgg = Entity.create( EntityType.EGG );
        entityEgg.setShooter( player );
        entityEgg.setLocation( player.getLocation().add( 0, player.getEyeHeight(), 0 ) );
        entityEgg.setVelocity( clickVector.multiply( 1.5f, 1.5f, 1.5f ), false );
        entityEgg.setYaw( player.getYaw() );
        entityEgg.setPitch( player.getPitch() );

        ProjectileLaunchEvent projectileLaunchEvent = new ProjectileLaunchEvent( entityEgg, ProjectileLaunchEvent.Cause.EGG );
        player.getServer().getPluginManager().callEvent( projectileLaunchEvent );
        if ( !projectileLaunchEvent.isCancelled() ) {
            this.updateItem( player, 1 );
            if ( player.getGameMode().equals( GameMode.SURVIVAL ) ) {
                player.getInventory().removeItem( ItemType.EGG, 1 );
            }
            entityEgg.spawn();
            player.getWorld().playSound( player.getLocation(), SoundEvent.THROW, -1, "minecraft:player", false, false );
            return true;
        }

        return false;
    }
}

File: JukeboxMC-Server/src/main/kotlin/org/jukeboxmc/server/item/behavior/ItemEgg.kt
package org.jukeboxmc.server.item.behavior

import org.cloudburstmc.protocol.bedrock.data.SoundEvent
import org.jukeboxmc.api.event.entity.ProjectileLaunchEvent
import org.jukeboxmc.api.item.Item
import org.jukeboxmc.api.item.ItemType
import org.jukeboxmc.api.math.Vector
import org.jukeboxmc.api.player.GameMode
import org.jukeboxmc.server.JukeboxServer
import org.jukeboxmc.server.entity.projectile.JukeboxEntityEgg
import org.jukeboxmc.server.item.JukeboxItem
import org.jukeboxmc.server.player.JukeboxPlayer

class ItemEgg(itemType: ItemType, countNetworkId: Boolean) : JukeboxItem(itemType, countNetworkId) {

    override fun useInAir(player: JukeboxPlayer, clickVector: Vector): Boolean {
        val entityEgg = JukeboxEntityEgg()
        entityEgg.setShooter(player)
        entityEgg.setLocation(player.getLocation().add(0f, player.getEyeHeight(), 0f))
        entityEgg.setMotion(clickVector.multiply(1.5f, 1.5f, 1.5f))

        val projectileLaunchEvent = ProjectileLaunchEvent(entityEgg, ProjectileLaunchEvent.Cause.EGG)
        JukeboxServer.getInstance().getPluginManager().callEvent(projectileLaunchEvent)
        if (projectileLaunchEvent.isCancelled()) return false

        if (player.getGameMode() == GameMode.SURVIVAL) {
            player.getInventory().removeItem(Item.Companion.create(ItemType.EGG), 1)
        }
        entityEgg.spawn()
        player.getWorld().playLevelSound(player.getLocation(), SoundEvent.THROW, -1, "minecraft:player")
        return true
    }

}